From 4b1fa178467d92c6279d62b415341fc393f42a08 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 12 Jan 2024 16:27:32 +0100
Subject: [PATCH 0438/1695] plat-stm32mp1: scmi: clock duty cycle

Some driver from non secure need to known the duty cycle value of
their clock.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1516577e8ec07ae3766c04fe6e675b4919bd05c4 plat-stm32mp1: scmi: clock duty cycle]
Change-Id: I6af2b3d5435c10d2e59dded71dc2501f1b970e4a
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/scmi_server.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/scmi_server.c b/core/arch/arm/plat-stm32mp1/scmi_server.c
index 992bdbbad..66e7fa9e3 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server.c
@@ -492,6 +492,25 @@ int32_t plat_scmi_clock_set_state(unsigned int channel_id, unsigned int scmi_id,
 	return SCMI_SUCCESS;
 }
 
+int32_t plat_scmi_clock_get_duty_cycle(unsigned int channel_id, unsigned int scmi_id,
+				       struct clk_duty *duty)
+{
+	struct stm32_scmi_clk *clock = find_clock(channel_id, scmi_id);
+	struct clk *clk_ref = NULL;
+	int ret = 0;
+
+	if (!clock || !stm32mp_nsec_can_access_clock(clock->clock_id))
+		return SCMI_DENIED;
+
+	clk_ref = stm32mp_rcc_clock_id_to_clk(clock->clock_id);
+
+	ret = clk_get_duty_cycle(clk_ref, duty);
+	if (ret)
+		return SCMI_NOT_SUPPORTED;
+
+	return SCMI_SUCCESS;
+}
+
 /*
  * Platform SCMI reset domains
  */
-- 
2.47.2

